### ISSUE TITLE
(6.0.7) `uniquejobs:{digest}:AVAILABLE` keys never expire

### ISSUE THREAD

**Issue by zvkemp** [mhenrixon/sidekiq-unique-jobs#359]
**Describe the bug**
`uniquejobs:{digest}:AVAILABLE` keys never expire. Given an unbounded set of unique arguments, this will fill up redis. 

**Expected behavior**
In version 6.0.6, these keys don't outlive the execution of their jobs.

**Current behavior**
`AVAILABLE` keys are not cleaned up (introduced in https://github.com/mhenrixon/sidekiq-unique-jobs/pull/354).  If there's a purpose to persisting them, it's not evident (at least in our use case)

**Worker class**

```ruby
class MyWorker
  include Sidekiq::Worker
  sidekiq_options lock: :until_executed
  def perform(args); end

  def self.unique_args(args)
    args
  end
end
```

**Comment by henrik** [comment]
+1. I just came here to report the same thing. Our Redis on one project ran out of space, and literally 98.6 % of the keys are "uniquejobs:…" :)

**Comment by jwg2s** [comment]
We're also seeing this - our total redis key count has been steadily rising since upgrading to v6.0.7

**Comment by henrik** [comment]
@jwg2s Deleting the keys immediately might mean the plugin doesn't do what it should, i.e. doesn't ensure jobs are unique. Though I don't know if these specific keys are used for that.

What we did was something similar but instead set an expiration on the keys, e.g. `$redis.expire(key, 8.hours)`. Our Redis on that server (a staging server) uses the volatile-lru strategy when hitting max memory, so it removes the least recently used key that has expiration. My hope is that this means it is less likely to remove keys that are being used to ensure jobs are unique.

**Comment by jwg2s** [comment]
@henrik good point - I've updated my script in my comment to use that strategy instead

**Comment by vitalinfo** [comment]
@jwg2s great workaround!

will be good to know why these keys are stuck in the redis?

**Comment by mhenrixon** [comment]
The available key becoming available is crucial for new jobs to be able to be scheduled. Them living on forever is most likely a terrible side-effect. Fix one bug (jobs can be scheduled properly) and cause another (redis grows out of proportion).

They could be created/added with an expiry. The problem is when using Luna is that it _might_ be that the `LPUSH` needs to be last.

**Comment by vitalinfo** [comment]
@mhenrixon what do you think, is it possible to fix, or we should leave with workaround that @jwg2s proposed?

**Comment by mhenrixon** [comment]
I'll try something tomorrow morning @vitalinfo. Will keep you guys posted

**Comment by mhenrixon** [comment]
Released with v6.0.8 @vitalinfo @jwg2s @henrik @zvkemp

**Comment by henrik** [comment]
@mhenrixon Thanks so much ❤️ Would you recommend we do anything to expire existing keys, or will those automatically be handled when the gem is updated?

**Comment by jwg2s** [comment]
@henrik I updated the script to use `5.seconds` as well, so it should match the new behavior found in v6.0.8.

**Comment by henrik** [comment]
Hm, that should be fine if locks aren’t needed for longer than 5 seconds. If one has much longer-running jobs then I guess there’s a risk that would stop locking those prematurely – assuming this key is important to keeping things locked. I think each person running it will need to figure out what expiry makes sense for them :)

I assume the lib only sets a 5 sec expiry after a job is done and the lock is no longer needed but I haven’t digested the code fully.

**Comment by blarralde** [comment]
Unfortunately that doesn't seem to have fixed it. 1 week after updating the gem there are 200k uniquejobs keys in our redis.

**Comment by blarralde** [comment]
Yes workers do restart a lot as we scale them up and down. They're scaled down only after jobs are done though so they should have time to clean up the queue?
